### ISSUE TITLE
Initialization loggerFactory limitations

### ISSUE THREAD

**Issue by DaviGia** [SteeltoeOSS/Steeltoe#675]
## Is your feature request related to a problem? Please describe.
Currently some `IHostBuilder` extensions methods (e.g. `AddKubernetesConfiguration()`, `AddConfigServer()`) accepts an optional `ILoggerFactory`.

Because this code is evaluated before the DynamicSerilogProvider is created, the loggerFactory can only be configured using a static serilog configuration, i.e. its log level cannot be changed at runtime and cannot be configured from appsettings.

``` cs
//create loggerFactory from a static configuration (not ideal)
ILoggerFactory initLogFactory = new LoggerFactory().AddSerilog(new LoggerConfiguration().MinimumLevel.Error().WriteTo.Console().CreateLogger());

return Host.CreateDefaultBuilder(args)
    .AddDynamicSerilog()
    .AddKubernetesConfiguration(null, initLogFactory) //initLogFactory is not configured with DynamicSerilog
    .AddConfigServer(initLogFactory) //initLogFactory is not configured with DynamicSerilog
    .AddDiscoveryClient()
    .AddAllActuators()
    .ConfigureWebHostDefaults(webBuilder => webBuilder.UseStartup<Startup>());
```

## Describe the solution you'd like
I would like to be able to pass loggerFactory in a lazy manner, so that I can build it when the DynamicSerilog is available. If that is not possible maybe those extension method may avoid to receive a loggerFactory as parameter, andthey could take it using DI instead (is it possible?).

## Describe alternatives you've considered
Use a static Serilog `LoggerConfiguration`.

**Comment by TimHess** [comment]
First off, thanks for the feedback, it's great to have in general. This is an enhancement we'd love to support, but unfortunately it is not _our_ limitation - see [Microsoft's guidance](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-5.0#log-during-host-construction) on logging during host construction. 

AFAIK what Steeltoe offers today is close to the best we could do. I think these methods likely need to stick with Microsoft's abstractions to avoid locking into a package like Serilog for those that don't want to use it. There might be some cases where we could accept an ILogger instead of ILoggerFactory, but there might be cases where we create multiple loggers from the factory and it seems simpler (IMO) to just pass around the factory. We would certainly be open to improvements provided by the community here, but I don't think the core team would be looking to work on this area any time soon.

**Comment by hananiel** [comment]
@DaviGia I am wondering what the use case for this is. The logFactory is only used in the build phase. So your app is not ready and available for you to change configuration dynamically. Can you please clarify? Is it that it cannot be configured from app settings?

**Comment by DaviGia** [comment]
Hi @hananiel,

the problem I'm facing is that if an error occurs during initialization phase of the microservice the error is printed using a logFactory attached to serilog that hasn't yet read the configuration from other sources (k8s resources, appsettings conf files, etc...). This is problematic in certain deployment environments (e.g. kubernetes) because the logging configuration may specify additional Serilog sinks (for example to push messages to ELK stack, that cannot be hardcoded in the code). My logging strategy is to push logs of all microservices in ELK to simplify log audit (all centralized in a single service).

The enhancement I have in mind is to find some way (if feasible, I don't know) to allow to specify an ILoggerFactory that can use DI to look for additional configuration (kubernetes resources, appsettings, config-server, etc..). Currently I haven't find a way to read Serilog configuration from IConfiguration before calling `Host.CreateDefaultBuilder(args)`. This is required because the Steeltoe methods requires the ILoggerFactory instance immediately. I think it could work if a `lazy variable`/`Func that consumes buildContext and produces a ILoggerFactory` is accepted (i.e. it can be evaluated later when the IConfiguration service is available, provided that a buildContext is available to retrieve the IConfiguration).

**Comment by hananiel** [comment]
@DaviGia To clarify, the requirement has to do with the ILoggerFactory respecting the available configuration and not with the dynamically changing the log level via actuators. If you have a sample with the other config sources you mention that will be great.

**Comment by DaviGia** [comment]
> the requirement has to do with the ILoggerFactory respecting the available configuration and not with the dynamically changing the log level via actuators

that is true, but I think it can be discussed here (if you don't think it's a good idea let me know and I will open another issue).

> If you have a sample with the other config sources you mention that will be great

I do not have a sample right now (but I can make it if you want, just tell me what you need). Currently I use Kubernetes and ConfigServer. You can use the snippet above.

**Comment by TimHess** [comment]
Restating one more time for clarity, we believe there are two separate issues in this issue:
1. A logger factory should use the most-current version of app configuration (handle config provider layering adding/overwriting details)

   * We could conceivably come up with a lazy construction approach
   * Right now you can do something like this:
      ```cs
      Host.CreateDefaultBuilder(args)
       .ConfigureAppConfiguration(builder => {
           // use all config providers added before <current> to configure dedicated logger
           builder.AddConfigServer(BuildLoggerFactory(builder.Build()));
           builder.AddKubernetes(BuildLoggerFactory(builder.Build()));
       });

      private static ILoggerFactory BuildLoggerFactory(IConfiguration config)
            => new LoggerFactory().AddSerilog(new LoggerConfiguration().ReadFrom.Configuration(config).MinimumLevel.Error().WriteTo.Console().CreateLogger());
      ```
   * I think ultimately this is nearly identical to what would happen with a lazy construction approach.
2.  The loggers used in config providers need to be compatible with dynamic logging

     * I think we can use an [`IHostedService`](https://docs.microsoft.com/aspnet/core/fundamentals/host/hosted-services) to swap out bootstrapped components with pieces from DI when they're all built - see first draft of such a service [here](https://github.com/SteeltoeOSS/Steeltoe/pull/737/files#diff-85384a7da7743f627abe4b48df78dfb90f888b8c54f9b7acf36d062002d3faf0)

**Comment by hananiel** [comment]
@DaviGia I have opened #780 to address this issue. When can try the PR with a sample this way, 1. The configured logger settings via any config source will be respected, 2. Dynamic logger settings will be activated and set via the logger actuator. : 

```csharp

   public static IHostBuilder CreateHostBuilder(string[] args) =>
            Host.CreateDefaultBuilder(args)
                .ConfigureWebHostDefaults(webBuilder =>
                {
                    webBuilder.UseStartup<Startup>();
                })
                .AddKubernetesConfiguration(null, GetLoggerFactory())
                .AddDynamicSerilog()
                .AddKubernetesActuators();

        private static ILoggerFactory GetLoggerFactory() => new LoggerFactory().AddSerilog(new LoggerConfiguration().MinimumLevel.Verbose().WriteTo.Console().CreateLogger());

```
